### PR TITLE
Dynamodb-enhanced [preview]: Updates table and index interfaces and r…

### PR DESCRIPTION
**Pull request by cenedhryn**
…efactors additional operations.

## Description
- Renames MappedIndex and MappedTable classes to be consistent with client interfaces
- Adds operation methods for Query, Scan and CreateTables on those interfaces
- Splits operations into request objects and internal operations
- Updates tests

## Motivation and Context
Improves discoverability by renaming the table and index interfaces to be consistent with the client interface naming, and by adding operation methods for createTable(), scan() and query(), as applicable. 

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)

## Checklist

- [x] I have read the **CONTRIBUTING** document
- [x] Local run of `mvn install` succeeds
- [x] My code follows the code style of this project
- [ ] My change requires a change to the Javadoc documentation
- [ ] I have updated the Javadoc documentation accordingly
- [x] I have read the **README** document
- [x] I have added tests to cover my changes
- [x] All new and existing tests passed
- [x] A short description of the change has been added to the **CHANGELOG**
- [ ] My change is to implement 1.11 parity feature and I have updated [LaunchChangelog](https://github.com/aws/aws-sdk-java-v2/blob/master/docs/LaunchChangelog.md)

## License

- [x] I confirm that this pull request can be released under the Apache 2 license

### CODE REVIEW

File: services-custom/dynamodb-enhanced/src/main/java/software/amazon/awssdk/extensions/dynamodb/mappingclient/model/GlobalSecondaryIndex.java [hunk 1]
```diff
@@ -25,17 +25,25 @@
...
+        this.projection = builder.projection;
+        this.provisionedThroughput = builder.provisionedThroughput;
     }
 
     public static GlobalSecondaryIndex create(String indexName,
                                               Projection projection,
                                               ProvisionedThroughput provisionedThroughput) {
 
-        return new GlobalSecondaryIndex(indexName, projection, provisionedThroughput);
+        return builder().indexName(indexName).projection(projection).provisionedThroughput(provisionedThroughput).build();
```
Comment (spfink) [inline services-custom/dynamodb-enhanced/src/main/java/software/amazon/awssdk/extensions/dynamodb/mappingclient/model/GlobalSecondaryIndex.java]: New line the different methods call for readability?

File: services-custom/dynamodb-enhanced/src/main/java/software/amazon/awssdk/extensions/dynamodb/mappingclient/model/GlobalSecondaryIndex.java [hunk 2]
```diff
@@ -25,17 +25,25 @@
...
+    }
+
+    public static Builder builder() {
+        return new Builder();
+    }
+
+    public Builder toBuilder() {
+        return new Builder().indexName(indexName).projection(projection).provisionedThroughput(provisionedThroughput);
```
Comment (spfink) [inline services-custom/dynamodb-enhanced/src/main/java/software/amazon/awssdk/extensions/dynamodb/mappingclient/model/GlobalSecondaryIndex.java]: Same as above

Comment (spfink) [inline services-custom/dynamodb-enhanced/src/main/java/software/amazon/awssdk/extensions/dynamodb/mappingclient/model/GlobalSecondaryIndex.java]: create() calls builder() while toBuilder does new Builder()

Should we make them consistent here and everywhere else?

File: services-custom/dynamodb-enhanced/src/test/java/software/amazon/awssdk/extensions/dynamodb/mappingclient/functionaltests/AsyncIndexQueryOperationTest.java [hunk 3]
```diff
@@ -0,0 +1,308 @@
...
+        mappedTable.createTable(CreateTableEnhancedRequest.builder()
+                                                          .provisionedThroughput(getDefaultProvisionedThroughput())
+                                                          .globalSecondaryIndices(
+                                                              GlobalSecondaryIndex.create(
+                                                                  "gsi_keys_only",
+                                                                  Projection.builder()
+                                                                            .projectionType(ProjectionType.KEYS_ONLY)
+                                                                            .build(),
+                                                                  getDefaultProvisionedThroughput()))
+                                                      .build())
```
Comment (spfink) [inline services-custom/dynamodb-enhanced/src/test/java/software/amazon/awssdk/extensions/dynamodb/mappingclient/functionaltests/AsyncIndexQueryOperationTest.java]: Where is this build hanging off of?

File: services-custom/dynamodb-enhanced/src/test/java/software/amazon/awssdk/extensions/dynamodb/mappingclient/functionaltests/IndexQueryOperationTest.java [hunk 4]
```diff
@@ -163,24 +163,25 @@ public int hashCode() {
...
+        mappedTable.createTable(CreateTableEnhancedRequest.builder()
+                                                          .provisionedThroughput(getDefaultProvisionedThroughput())
+                                                          .globalSecondaryIndices(
+                                                              GlobalSecondaryIndex.create(
+                                                                  "gsi_keys_only",
+                                                                  Projection.builder()
+                                                                            .projectionType(ProjectionType.KEYS_ONLY)
+                                                                            .build(),
+                                                                  getDefaultProvisionedThroughput()))
+                                                      .build());
```
Comment (spfink) [inline services-custom/dynamodb-enhanced/src/test/java/software/amazon/awssdk/extensions/dynamodb/mappingclient/functionaltests/IndexQueryOperationTest.java]: Where is this hanging from?

File: services-custom/dynamodb-enhanced/src/main/java/software/amazon/awssdk/extensions/dynamodb/mappingclient/DynamoDbAsyncIndex.java [hunk 5]
```diff
@@ -27,28 +27,15 @@
...
-     * context of the specific table and secondary index this object is linked to.
-     *
-     * @param operationToPerform The operation to be performed in the context of the secondary index.
-     * @param <R> The expected return type from the operation. This is typically inferred by the compiler.
-     * @return An {@link SdkPublisher} that will publish successive pages of result data items to any subscriber with
-     * demand for them.
-     */
-    <R> SdkPublisher<R> execute(PaginatedIndexOperation<T, ?, ?, R> operationToPerform);
+    default SdkPublisher<Page<T>> query(QueryEnhancedRequest request) {
+        throw new UnsupportedOperationException();
```
Comment (spfink) [inline services-custom/dynamodb-enhanced/src/main/java/software/amazon/awssdk/extensions/dynamodb/mappingclient/DynamoDbAsyncIndex.java]: Why the need for defaults here?

Comment (cenedhryn) [reply services-custom/dynamodb-enhanced/src/main/java/software/amazon/awssdk/extensions/dynamodb/mappingclient/DynamoDbAsyncIndex.java]: For supporting new operations in the future, I think, and now for consistency. It's in the design doc.

**Pull request merged**